### PR TITLE
New producers and filters for the Phase 2 Muon HLT

### PR DESCRIPTION
**Pull request by khaosmos93**
#### PR description:
This PR adds the following modules:
- a new L2 muon seed generator from L1TkMuon
- a new L3 muon filter with L1TkMuon
- a new dR filter for MuonRef of L1TkMuon

N.B. To compile it, #32571 should be merged prior to this PR

#### PR validation:
matrix tests (after merging #32571)

### CODE REVIEW

**Comment by fwyzard** [comment]
code-checks

**Comment by fwyzard** [comment]
urgent

**Comment by fwyzard** [comment]
please test with #32571

**Comment by perrotta** [comment]
code-checks
(now that #32571 is merged in the master, since CMSSW_11_3_X_2021-01-20-2300, the clang build errors should disappear)

**Comment by perrotta** [comment]
please test

File: HLTrigger/Muon/plugins/HLTMuonTrkL1TkMuFilter.cc [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  std::vector<unsigned int> filteredMuons;
+  for (unsigned int i = 0; i < muons->size(); ++i) {
+    const reco::Muon& muon(muons->at(i));
+    // check for dR match to L1 muons
+    if (check_l1match) {
+      bool matchl1 = false;
+      for (auto l1cand = vl1cands_begin; l1cand != vl1cands_end; ++l1cand) {
+        if (deltaR(muon, **l1cand) < 0.3) {
```
Comment (fwyzard) [inline HLTrigger/Muon/plugins/HLTMuonTrkL1TkMuFilter.cc:96]: this could be simplified as
```suggestion
      for (auto const& l1cand: vl1cands) {
        if (deltaR(muon, *l1cand) < 0.3) {
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+class MuonServiceProxy;
+class MeasurementEstimator;
+class TrajectorySeed;
+class TrajectoryStateOnSurface;
+
+namespace edm {
+  class ParameterSet;
+  class Event;
+  class EventSetup;
+}  // namespace edm
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:45]: rather than forward declaring these classes directly, it is better praxtice to include their forward declarations header:
```suggestion
#include "FWCore/Framework/interface/Frameworkfwd.h"
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  const double theMinPtL1TkBarrel;
+  const bool useOfflineSeed;
+  const bool useUnassociatedL1;
+  std::vector<double> matchingDR;
+  std::vector<double> etaBins;
+
+  /// the event setup proxy, it takes care the services update
+  MuonServiceProxy *theService;
+
+  MeasurementEstimator *theEstimator;
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:80]: these should all get an underscore at the end of their name:
```suggestion
  const double theL1MinPt_;
  const double theL1MaxEta_;
  const double theMinPtBarrel_;
  const double theMinPtEndcap_;
  const double theMinPL1Tk_;
  const double theMinPtL1TkBarre_l;
  const bool useOfflineSeed_;
  const bool useUnassociatedL1_;
  std::vector<double> matchingDR_;
  std::vector<double> etaBins_;

  /// the event setup proxy, it takes care the services update
  MuonServiceProxy *theService_;

  MeasurementEstimator *theEstimator_;
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  /// Destructor
+  ~L2MuonSeedGeneratorFromL1TkMu() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+  void produce(edm::Event &, const edm::EventSetup &) override;
+
+private:
+  edm::InputTag theSource;
+  edm::InputTag theOfflineSeedLabel;
+  std::string thePropagatorName;
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:61]: these should get an underscore at the end of their name:
```suggestion
  edm::InputTag theSource_;
  edm::InputTag theOfflineSeedLabel_;
  std::string thePropagatorName_;
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+      theL1MaxEta(iConfig.getParameter<double>("L1MaxEta")),
+      theMinPtBarrel(iConfig.getParameter<double>("SetMinPtBarrelTo")),
+      theMinPtEndcap(iConfig.getParameter<double>("SetMinPtEndcapTo")),
+      theMinPL1Tk(iConfig.getParameter<double>("MinPL1Tk")),
+      theMinPtL1TkBarrel(iConfig.getParameter<double>("MinPtL1TkBarrel")),
+      useOfflineSeed(iConfig.getUntrackedParameter<bool>("UseOfflineSeed", false)),
+      useUnassociatedL1(iConfig.getParameter<bool>("UseUnassociatedL1")),
+      matchingDR(iConfig.getParameter<std::vector<double>>("MatchDR")),
+      etaBins(iConfig.getParameter<std::vector<double>>("EtaMatchingBins")) {
+  muCollToken_ = consumes<l1t::TkMuonCollection>(theSource);
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:54]: this sholud also be initialised in the member initialisation, above:
```suggestion
  muCollToken_(consumes<l1t::TkMuonCollection>(theSource))
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h (hunk 3 above)
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:80]: These should be `std::unique_ptr<...>` rather than raw pointers:
```suggestion
  /// the event setup proxy, it takes care the services update
  std::unique_ptr<MuonServiceProxy> theService_;
  std::unique_ptr<MeasurementEstimator> theEstimator_;
```

Or, they could simple be used by value:
```suggestion
  /// the event setup proxy, it takes care the services update
  MuonServiceProxy theService_;
  MeasurementEstimator theEstimator_;
  ```
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+  edm::Handle<edm::View<TrajectorySeed>> offlineSeedHandle;
+  vector<int> offlineSeedMap;
+  if (useOfflineSeed) {
+    iEvent.getByToken(offlineSeedToken_, offlineSeedHandle);
+    offlineSeedMap = vector<int>(offlineSeedHandle->size(), 0);
+  }
+
+  for (auto ittkmu = muColl->begin(); ittkmu != muColl->end(); ittkmu++) {
+    // L1 tracker track
+    auto it = ittkmu->trkPtr();
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:130]: this could be simplified with a range-based loop:
```suggestion
  for (auto const& tkmu : muColl)) {
    // L1 tracker track
    auto it = tkmu.trkPtr();
```

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+
+      // MB2
+      theid = DTChamberId(0, 2, 0);
+      detLayer = theService->detLayerGeometry()->idToLayer(theid);
+      LogDebug(metname) << "L2 Layer: " << debug.dumpLayer(detLayer);
+
+      const BoundSurface *sur = &(detLayer->surface());
+      const BoundCylinder *bc = dynamic_cast<const BoundCylinder *>(sur);
+
+      radius = fabs(bc->radius() / sin(theta));
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:212]: everywhere, please use `std::abs()` instead of `fabs`

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+      LogDebug(metname) << "radius " << radius;
+
+      if (pt < theMinPtBarrel)
+        pt = theMinPtBarrel;
+    } else {
+      LogDebug(metname) << "The seed is in the endcap";
+
+      // ME2
+      theid = theta < Geom::pi() / 2. ? CSCDetId(1, 2, 0, 0, 0) : CSCDetId(2, 2, 0, 0, 0);
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:222]: it might be better to use `M_PI_2` instead of `Geom::pi() / 2.`, but I don't know if there are any guidelines about this.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+    auto p3 = it->momentum();
+    float tk_pt = p3.perp();
+    float tk_p = p3.mag();
+    float tk_eta = p3.eta();
+    float tk_aeta = std::abs(tk_eta);
+    float tk_phi = p3.phi();
+    float tk_q = it->rInv() > 0 ? 1. : -1.;
+    float tk_z = it->POCA().z();
+
+    if (tk_p < theMinPL1Tk)
```
Comment (fwyzard) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:142]: throughout this function, could you clarify what constants like `1.1`, `550.`, `850.`, `1.464`, *etc.* mean, and where they come from ?

IMHO adding a comment the first time they are used should be enough; defining all the named constants in a single place and using them consistently would be even clearer.

**Review by fwyzard — APPROVED** [review]
There are some trivial code issues that should be fixed.

However, I think we really need this in CMSSW_11_1_7, so I would suggest merging as is, and following up in the next days.

**Comment by fwyzard** [comment]
+hlt

**Comment by silviodonato** [comment]
Given that #32700 is needed for the upcoming re-HLT, I agree that it is better to merge the PR as it is in order to have the new modules in 11_1_7 instead of merging the PR manually.

**Comment by jpata** [comment]
(reco will have a chat about all PRs today at 5pm, we will reply here after that)

File: HLTrigger/HLTfilters/plugins/HLT2L1TkMuonL1TkMuonMuRefDR.cc [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+  float muRef1_eta = 0.;
+  float muRef1_phi = 0.;
+  if (r1->muonDetector() != 3) {
+    if (r1->muRef().isNull())
+      return false;
+
+    muRef1_eta = r1->muRef()->hwEta() * 0.010875;
+    muRef1_phi = static_cast<float>(l1t::MicroGMTConfiguration::calcGlobalPhi(
+        r1->muRef()->hwPhi(), r1->muRef()->trackFinderType(), r1->muRef()->processor()));
+    muRef1_phi = muRef1_phi * 2. * M_PI / 576.;
```
[the remark this answers is not in the document]
Comment (fwyzard) [reply HLTrigger/HLTfilters/plugins/HLT2L1TkMuonL1TkMuonMuRefDR.cc:116]: My suggestions would be to declare the same constants in this plugin, and use the constants instead of the bare numbers.

A following step could be to move the constants to a public header file under `L1Trigger/L1TTrackMatch` and reuse them from there - but MO that's beyond the scope of this PR, and should be done by L1T.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
+#ifndef RecoMuon_L2MuonSeedGenerator_L2MuonSeedGeneratorFromL1TkMu_H
+#define RecoMuon_L2MuonSeedGenerator_L2MuonSeedGeneratorFromL1TkMu_H
```
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:2]: by now we require that .h files for plugins are not made and only .cc is used.
Please merge this into .h

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h (hunk 3 above)
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:80]: can these be smart pointers?

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+                             ->idToDet(offseed->startingState().detId())
+                             ->surface()
+                             .toGlobal(offseed->startingState().parameters().position());
+    GlobalVector glbMom = theService->trackingGeometry()
+                              ->idToDet(offseed->startingState().detId())
+                              ->surface()
+                              .toGlobal(offseed->startingState().parameters().momentum());
+
+    // Preliminary check
+    double preDr = deltaR(newTsos.globalPosition().eta(), newTsos.globalPosition().phi(), glbPos.eta(), glbPos.phi());
```
Comment (perrotta) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:365]: you can use `deltaR2` instead, with better performance

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+    if (offseedTsos.isValid()) {
+      LogDebug(metlabel) << "Offline seed info after propagation to L1 layer:" << std::endl;
+      LogDebug(metlabel) << "pos: (r=" << offseedTsos.globalPosition().mag()
+                         << ", phi=" << offseedTsos.globalPosition().phi()
+                         << ", eta=" << offseedTsos.globalPosition().eta() << ")" << std::endl;
+      LogDebug(metlabel) << "mom: (q*pt=" << offseedTsos.charge() * offseedTsos.globalMomentum().perp()
+                         << ", phi=" << offseedTsos.globalMomentum().phi()
+                         << ", eta=" << offseedTsos.globalMomentum().eta() << ")" << std::endl
+                         << std::endl;
+      double newDr = deltaR(newTsos.globalPosition().eta(),
```
Comment (perrotta) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:391]: `deltaR2`

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h (hunk 4 above)
Comment (slava77) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.h:61]: > these should get an underscore at the end of their name:

`the` would better be dropped in this case

File: HLTrigger/HLTfilters/plugins/HLT2L1TkMuonL1TkMuonMuRefDR.cc [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+  } else {
+    if (r2->emtfTrk().isNull())
+      return false;
+
+    muRef2_eta = r2->emtfTrk()->Eta();
+    muRef2_phi = angle_units::operators::convertDegToRad(r2->emtfTrk()->Phi_glob());
+  }
+  muRef2_phi = reco::reduceRange(muRef2_phi);
+
+  if (reco::deltaR(muRef1_eta, muRef1_phi, muRef2_eta, muRef2_phi) > minDR_)
```
Comment (perrotta) [inline HLTrigger/HLTfilters/plugins/HLT2L1TkMuonL1TkMuonMuRefDR.cc:145]: `deltaR2`

File: HLTrigger/Muon/plugins/HLTMuonTrkL1TkMuFilter.cc (hunk 1 above)
Comment (perrotta) [reply HLTrigger/Muon/plugins/HLTMuonTrkL1TkMuFilter.cc:96]: `deltaR2`

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+    float resPhi = tk_phi - 1.464 * tk_q * cosh(1.7) / cosh(etaProp) / tk_pt * dzCorrPhi - M_PI / 144.;
+    resPhi = reco::reduceRange(resPhi);
+
+    float pt = tk_pt;  //not corrected for eloss
+    float eta = tk_eta + deta;
+    float theta = 2 * atan(exp(-eta));
+    float phi = resPhi;
+    int charge = it->rInv() > 0 ? 1 : -1;
+
+    bool barrel = tk_aeta < 1.1 ? true : false;
```
Comment (perrotta) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:176]:
```suggestion
    bool barrel = tk_aeta < 1.1;
```
And you can define it at the beginning, thus avoiding the several previous checks `tk_aeta < 1.1`

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+    offlineSeedMap = vector<int>(offlineSeedHandle->size(), 0);
+  }
+
+  for (auto ittkmu = muColl->begin(); ittkmu != muColl->end(); ittkmu++) {
+    // L1 tracker track
+    auto it = ittkmu->trkPtr();
+
+    // propagate to GMT
+    auto p3 = it->momentum();
+    float tk_pt = p3.perp();
```
Comment (perrotta) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:134]: The check on minimum pt can be applied here, just to avoid useless computations if under threshold

**Comment by silviodonato** [comment]
merge
@khaosmos93 please prepare a follow-up pull request with the implementation of the pending comments (#32699)

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+          if (newTSOS.isValid()) {
+            LogDebug(metname) << "pos: (r=" << newTSOS.globalPosition().mag()
+                              << ", phi=" << newTSOS.globalPosition().phi()
+                              << ", eta=" << newTSOS.globalPosition().eta() << ")";
+            LogDebug(metname) << "mom: (q*pt=" << newTSOS.charge() * newTSOS.globalMomentum().perp()
+                              << ", phi=" << newTSOS.globalMomentum().phi()
+                              << ", eta=" << newTSOS.globalMomentum().eta() << ")";
+
+            const TrajectorySeed *assoOffseed =
+                associateOfflineSeedToL1(offlineSeedHandle, offlineSeedMap, newTSOS, dRcone);
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:305]: please use a smart pointer.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+    LogDebug(metname) << "eta = " << eta;
+    LogDebug(metname) << "theta = " << theta << " rad";
+    LogDebug(metname) << "phi = " << phi << " rad";
+    LogDebug(metname) << "charge = " << charge;
+    LogDebug(metname) << "In Barrel? = " << barrel;
+
+    // Update the services
+    theService->update(iSetup);
+
+    const DetLayer *detLayer = nullptr;
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:192]: please use a smart pointer

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+            const TrajectorySeed *assoOffseed =
+                associateOfflineSeedToL1(offlineSeedHandle, offlineSeedMap, newTSOS, dRcone);
+
+            if (assoOffseed != nullptr) {
+              PTrajectoryStateOnDet const &seedTSOS = assoOffseed->startingState();
+              for (auto const &recHit : assoOffseed->recHits()) {
+                container.push_back(recHit);
+              }
+              auto dummyRef = edm::Ref<MuonBxCollection>();
+              output->push_back(L2MuonTrajectorySeed(seedTSOS, container, alongMomentum, dummyRef));
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:313]: emplace_back

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+            } else {
+              if (useUnassociatedL1) {
+                // convert the TSOS into a PTSOD
+                PTrajectoryStateOnDet const &seedTSOS =
+                    trajectoryStateTransform::persistentState(newTSOS, newTSOSDet->geographicalId().rawId());
+                auto dummyRef = edm::Ref<MuonBxCollection>();
+                output->push_back(L2MuonTrajectorySeed(seedTSOS, container, alongMomentum, dummyRef));
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:320]: emplace_back

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+              }
+            }
+          }
+        }
+      } else {
+        // convert the TSOS into a PTSOD
+        PTrajectoryStateOnDet const &seedTSOS = trajectoryStateTransform::persistentState(tsos, theid.rawId());
+        auto dummyRef = edm::Ref<MuonBxCollection>();
+        output->push_back(L2MuonTrajectorySeed(seedTSOS, container, alongMomentum, dummyRef));
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:329]: emplace_back

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+const TrajectorySeed *L2MuonSeedGeneratorFromL1TkMu::associateOfflineSeedToL1(
+    edm::Handle<edm::View<TrajectorySeed>> &offseeds,
+    std::vector<int> &offseedMap,
+    TrajectoryStateOnSurface &newTsos,
+    double dRcone) {
+  const std::string metlabel = "Muon|RecoMuon|L2MuonSeedGeneratorFromL1TkMu";
+  MuonPatternRecoDumper debugtmp;
+
+  edm::View<TrajectorySeed>::const_iterator offseed, endOffseed = offseeds->end();
+  const TrajectorySeed *selOffseed = nullptr;
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:347]: please use a smart pointer

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+      if (useOfflineSeed) {
+        // Get the compatible dets on the layer
+        std::vector<pair<const GeomDet *, TrajectoryStateOnSurface>> detsWithStates =
+            detLayer->compatibleDets(tsos, *theService->propagator(thePropagatorName), *theEstimator);
+
+        if (detsWithStates.empty() && barrel) {
+          // Fallback solution using ME2, try again to propagate but using ME2 as reference
+          DetId fallback_id;
+          theta < Geom::pi() / 2. ? fallback_id = CSCDetId(1, 2, 0, 0, 0) : fallback_id = CSCDetId(2, 2, 0, 0, 0);
+          const DetLayer *ME2DetLayer = theService->detLayerGeometry()->idToLayer(fallback_id);
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:283]: please use a smart pointer

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+
+          tsos = theService->propagator(thePropagatorName)->propagate(state, ME2DetLayer->surface());
+          detsWithStates = ME2DetLayer->compatibleDets(tsos, *theService->propagator(thePropagatorName), *theEstimator);
+        }
+
+        if (!detsWithStates.empty()) {
+          TrajectoryStateOnSurface newTSOS = detsWithStates.front().second;
+          const GeomDet *newTSOSDet = detsWithStates.front().first;
```
Comment (jpata) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:291]: please use a smart pointer

**Review by jpata — COMMENTED** [review]
hopefully this is helpful

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 17 above)
Comment (fwyzard) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:305]: no

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 18 above)
Comment (fwyzard) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:192]: actually, no, don't

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 22 above)
Comment (fwyzard) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:347]: actually, no, don't

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 23 above)
Comment (fwyzard) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:283]: actually, no, don't

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 24 above)
Comment (fwyzard) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:291]: actually, no, don't

**Comment by fwyzard** [comment]
@jpata, smart pointers should only be used to manage the ownership or lifetime of the pointed-to objects.
In this case, they are used to find and return to the caller objects that are owned by other data products in the Event or by conditions in the EventSetup: a raw `const` pointer (or a `const &` if it can be initialised contextually) is the right thing to use in this case.

**Comment by slava77** [comment]
> merge
> @khaosmos93 please prepare a follow-up pull request with the implementation of the pending comments (#32699)

@silviodonato 
for the future cases of extreme urgency to add a feature to a production release, can we instead use an exception to the "merge in master first" rule instead of ignoring the review process?
This has been done in the past and worked somewhat OK.

**Comment by perrotta** [comment]
> merge
> @khaosmos93 please prepare a follow-up pull request with the implementation of the pending comments (#32699)

@silviodonato Suggestion: wouldn't it be better to fully revert the integration of this PR in the master, and make another pull request with the same branch (which I see @khaosmos93 is already modifying)? It will be certainly easier to review a pull request which has all new/modified lines visible in the github diff, rather than one which only displays the newly modified lines of code, where the whole picture can be lost.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc (hunk 17 above)
Comment (jpata) [reply RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc:305]: if there's a bug in `associateOfflineSeedToL1` this can return a null pointer. In other new reco code, we have asked for stricter memory management.

**Comment by jpata** [comment]
> @jpata, smart pointers should only be used to manage the ownership or lifetime of the pointed-to objects.

Thanks for your insight and I agree in some cases when dealing with legacy geometry stuff it can be OK.

But if there's a function such as the one here: https://github.com/cms-sw/cmssw/pull/32699/files#diff-3dd45865ad219c6418f72b42cc3ce1f6a39f44b61aacb5239b164b1416c8543eR338
```C++
const TrajectorySeed *assoOffseed = nullptr;
...some logic...
return assoOffseed;
```
then there's always a danger of a bug slipping in, if not today, then perhaps later when a developer that is less than perfect touches tthis function, which cannot be excluded.

**Comment by silviodonato** [comment]
> > merge
> > @khaosmos93 please prepare a follow-up pull request with the implementation of the pending comments (#32699)
> 
> @silviodonato
> for the future cases of extreme urgency to add a feature to a production release, can we instead use an exception to the "merge in master first" rule instead of ignoring the review process?
> This has been done in the past and worked somewhat OK.

I agree, it would have been better to merge only the 11_1_X PR, keeping this PR open. Next time I will follow this suggestion

**Comment by fwyzard** [comment]
I think you are confusing what a smart pointer does ?

A smart pointer (like `std::shared_ptr` and `std::unique_ptr`) manages the lifetime of the pointed-to object.
And that's all.
It never give any guarantee that it is not "nullptr".

A function like
```c++
const TrajectorySeed *L2MuonSeedGeneratorFromL1TkMu::associateOfflineSeedToL1(...) {
  const TrajectorySeed *selOffseed = nullptr;
...some nontrivial logic...
  return selOffseed;
}
```
never manages the lifetime and ownership of any object.

Changing it to use a smart pointer (not sure how, but I assume you mean something like this):
```c++
std::shared_ptr<TrajectorySeed> L2MuonSeedGeneratorFromL1TkMu::associateOfflineSeedToL1(...) {
  std::shared_ptr<TrajectorySeed> selOffseed;
...some nontrivial logic where selOffseed is assigned a pointer to an existing object...
  return selOffseed;
}
```
does nothing to prevent it from returning a `std::shared_ptr`-to-`nulllptr`: if there is a bug affecting the original version, the same bug will affect it in the exact same way.

On the other, how are you populating the smart pointer ?
If you fill it with a copy of the original object, you are wasting memory and time without achieving anything useful.
If you fill it to a pointer to the original object (you need to `const_cast` away constness to do it) you have introduced a bug: when the smart pointer goes out of scope it will (try to) delete the object stored in the event - resulting in a crash immediately (if one is lucky) or the next time the same object is accessed (if one is unlucky).

Bottom line: *do not* use a smart pointer here, and in all instances where one is *not* managing the ownership and lifetime of the underlying object.

**Comment by fwyzard** [comment]
From the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-return-ptr):

> ### F.42: Return a T* to indicate a position (only)
> #### Reason
> 
> That’s what pointers are good for. Returning a T* to transfer ownership is a misuse.
> #### Example
> ```
> Node* find(Node* t, const string& s)  // find s in a binary tree of Nodes
> {
>     if (!t || t->name == s) return t;
>     if ((auto p = find(t->left, s))) return p;
>     if ((auto p = find(t->right, s))) return p;
>     return nullptr;
> }
> ```
> If it isn’t the `nullptr`, the pointer returned by find indicates a `Node` holding `s`. Importantly, that does not imply a transfer of ownership of the pointed-to object to the caller.
> 
> #### Note
> Positions can also be transferred by iterators, indices, and references. A reference is often a superior alternative to a pointer if there is no need to use `nullptr` or if the object referred to should not change.

**Comment by jpata** [comment]
@fwyzard thank you for the description.

Then let me rephrase my comment: I would like to ask to find a way to avoid the possibility (verified at compile time) of returning `nullptr` from this nontrivial function `associateOfflineSeedToL1`.

This is simply because I have not seen new code come in recently in the reco areas where this pattern with C-style pointers, null-initialized pointers is used.

**Comment by jpata** [comment]
@khaosmos93 from my point of view, it's also fine to do as @silviodonato has suggested and follow up with a second PR on master very shortly that fixes the review items pointed out in this thread.

It would be good if this can happen within a well-defined timeline that is committed to in advance, to avoid it slipping through the cracks.

The backport fix should be made when the master PR is merged.

**Comment by khaosmos93** [comment]
Hello everyone, Thank you very much for the comments.
I made a new PR #32713 addressing them.

**Comment by jpata** [comment]
@khaosmos93 thank you for the quick follow up. I will post review comments for both PRs here in an attempt to keep everything in one.

I still see a few things unaddressed, focusing only on code in the reco areas, without having compiled or run the code:
- forward defs to be removed https://github.com/cms-sw/cmssw/pull/32713/files#diff-70b2a719aac32ca979c0ac376047b49c9ed92e9098d0eca81fb05ea80b6ae16dR54-R57
- empty destructor may not be needed: https://github.com/cms-sw/cmssw/pull/32713/files#diff-70b2a719aac32ca979c0ac376047b49c9ed92e9098d0eca81fb05ea80b6ae16dR149
- confirm if it's possible to change the code to e.g. return a reference, rather than a C-style pointer which may be null: https://github.com/cms-sw/cmssw/pull/32699/files#diff-3dd45865ad219c6418f72b42cc3ce1f6a39f44b61aacb5239b164b1416c8543eR338

**Comment by fwyzard** [comment]
> Then let me rephrase my comment: I would like to ask to find a way to avoid the possibility (verified at compile time) of returning `nullptr` from this nontrivial function `associateOfflineSeedToL1`.

Then, I think what you are looking for is simply a reference ?

However in this case it would not work: the function needs to return `nullptr` is no matching `TrajectorySeed` is found.
In fact, the code that calls it **explicitly** checks for that condition:
```c++
             if (assoOffseed != nullptr) { 
               // do something with assoOffseed
             } else {
               // handle the case when no match is found
             }
``` 

Which is exactly how the search algorithms in the C++ standard library are built: they return an iterator to the object being searched for, or an invalid iterator is nothing is found.


> This is simply because I have not seen new code come in recently in the reco areas where this pattern with C-style, null-initialized pointers is used.

Then you can be grateful that you have learned something useful today.
When that happens to me, I know I usually am.

**Comment by jfernan2** [comment]
+1
